finality-aleph/src/party.rs

```diff
@@ -1,79 +1,392 @@
+            select_chain,
+            authority,
+            finalization_proposals_rx,
+            spawn_handle: spawn_handle.into(),
+            sessions: HashMap::new(),
+            phantom: PhantomData,
+        }
+    }
+
+    async fn run(mut self) {
+        // Prepare and start the network
+        let network = ConsensusNetwork::new(self.network.clone(), "/cardinals/aleph/1");
+        let session_manager = network.session_manager();
+
+        let task = async move { network.run().await };
+        self.spawn_handle.0.spawn("aleph/network", task);
+        debug!(target: "afa", "Consensus network has started.");
+
+        let (new_receivers_tx, new_receivers_rx) = mpsc::unbounded();
+        let mut proposition_select =
+            ProposalSelect::<OrderedBatch<B::Hash>>::new(new_receivers_rx).fuse();
+
+        let mut waiting_blocks = HashMap::<u32, Vec<B::Hash>>::new();
+        let max_len = 100;
+
+        let genesis_session = match self
+            .client
+            .runtime_api()
+            .current_session(&BlockId::Number(0.into()))
+        {
+            Ok(session) => session,
+            _ => {
+                error!(target: "afa", "No session found for current block #{}", 0);
+                return;
+            }
+        };
+
+        // Start new session if we are in the authority set.
+        let (instance, proposition_rx) = create_session(
+            self.authority.clone(),
+            self.auth_keystore.clone(),
+            self.select_chain.clone(),
+            self.spawn_handle.clone(),
+            genesis_session,
+            &session_manager,
+        );
+        self.sessions.insert(0, instance);
+        if let Some(proposition_rx) = proposition_rx {
+            new_receivers_tx
+                .unbounded_send((0, proposition_rx))
+                .expect("Sending channel should succeed");
+        }
+
+        for curr_id in 0.. {
+            // Stopping block is the last before the new session kick ins
+            let current_stop_h = self
+                .sessions
+                .get(&curr_id)
+                .expect("The current session should be known already")
+                .session
+                .stop_h;
+
+            let handle_proposal = |this: &mut Self, h: B::Hash| {
+                if let Some(reduced) = reduce_block_up_to(&this.client, h, current_stop_h) {
+                    if check_extends_finalized(&this.client, reduced) {
+                        finalize_block_as_authority(&this.client, reduced, &this.auth_keystore);
+                    }
+                }
+            };
+
+            if let Some(hashes) = waiting_blocks.get(&curr_id) {
+                for hash in hashes {
+                    handle_proposal(&mut self, *hash);
+                }
+            }
 
-    debug!(target: "afa", "Consensus party has stopped");
+            waiting_blocks.remove(&curr_id);
+
+            while self.client.info().finalized_number < current_stop_h {
```

Comment:
It is not clear to me whether we can rely on `client.info().finalized_number` being up-to-date -- what I mean is that client is perhaps an independent thread which might be busy when we call finalize_block() and thus not update finalized_number right away? Not sure about that, but one can probably ignore this issue by replacing `client.info().finalized_number` with a local variable that maintains this height (the assumption here is that party is the only place where we finalize blocks, so we should be able to update this variable suitably).

Reply:
I dont like the local variable which we will maintain here by updating height because we can't say if something was really finalized - I guess after all the checks it is highly unlikely but still possible. I think the client.info won't fall behind because we are blocking on the finalization in the `finalize_block` function and only after the operation ends we return to this loop here so the state of the client/backend should already be updated.

Reply:
Ok, let's leave it as it is then. The point is though that whether "something was really finalized" is determined by this exact piece of code. So in other words: this piece of code is instructing the client to mark some blocks as finalized, there is no such thing as some external finalization event that we are not aware of here.